### PR TITLE
chore(go): runtime-wide type registration

### LINKED ISSUE #2561: Go runtime type registration
## :rocket: Feature Request

We need to re-work the type handling, moving way from the `impMap` passed to the runtimes `Invoke`, `StaticInvoke`, `Get` and `StaticGet` calls, and instead have a client-wide type registry.

This is necessary to enable returning the correct dynamic type in several occasions:
- Declared type is `ParentClass`, dynamic value is `ChildClass extends ParentClass`
- Returning values that implement a collection of interfaces
- ...

This may also be necessary to be able to correctly handle enum values; and to be able to re-hydrate structs correctly.

### PR DESCRIPTION
**Pull request by RomainMuller**
The previous convention of passing of `implementationMap` values at the
call sites for value-returning calls (`Get`, `Invoke`, and their static
counterparts) did not provide sufficient information to properly convert
certain values (JavaScript returning a descendant of the declared return
type, decoding enum members, ability to represent values that implement
multiple interfaces in JavaScript, ...).

In order to address this shortcomings, this mechanism was replaced with
a new `func init(){}` block emitted in each package, which is
responsible for registering the generated types with the runtime
library. This allows the `client` to determine the correct go type to
use when converting arbitrary values returned from JavaScript, as it has
visibility on the complete set of available types, by fully qualified
name.

This minimally changes the conversion of JavaScript values into go
values, so as to minimize the footprit of this change. Of course,
subsequent effort will be made to fully leverage the potential of the
extended information (i.e: correctly processing enum members, ...).

Fixes #2561

---

By submitting this pull request, I confirm that my contribution is made under the terms of the [Apache 2.0 license].

[Apache 2.0 license]: https://www.apache.org/licenses/LICENSE-2.0

### CODE REVIEW

File: packages/@jsii/go-runtime/jsii-runtime-go/type-registry.go
Comment (eladb) [inline packages/@jsii/go-runtime/jsii-runtime-go/type-registry.go:107]: Should this return an `Error` instead of a boolean?

File: packages/jsii-pacmak/lib/targets/go/package.ts
Comment (eladb) [inline packages/jsii-pacmak/lib/targets/go/package.ts:111]: Maybe use `path.basename(this.file, '.go')` instead?

File: packages/jsii-pacmak/lib/targets/go/package.ts
Comment (eladb) [inline packages/jsii-pacmak/lib/targets/go/package.ts:110]: Add some commentary on what is this function (and consider extracting to a separate method)

File: packages/jsii-pacmak/lib/targets/go/package.ts
Comment (eladb) [inline packages/jsii-pacmak/lib/targets/go/package.ts:119]: This seems like a reporting pattern in the Go code gen, maybe extract to a helper method on `Package`?

```ts
this.emitImports('reflect', `${JSII_RT_ALIAS} "${JSII_RT_MODULE_NAME}"`);
```

Or:

```ts
this.emitGoClause('imports', [ ... ]);
```

File: packages/jsii-pacmak/test/generated-code/__snapshots__/target-go.test.ts.snap
Comment (eladb) [inline packages/jsii-pacmak/test/generated-code/__snapshots__/target-go.test.ts.snap:211]: Is that the idiomatic way in Go to obtain a reflection type?

**Review by eladb — COMMENTED** [review]
LGTM, I'll leave it to @MrArnoldPalmer to approve

File: packages/@jsii/go-runtime/build-tools/gen-calc.ts
Comment (RomainMuller) [reply packages/@jsii/go-runtime/build-tools/gen-calc.ts:18]: Could not get it to build locally without specifying this... The `go build` would always fail to resolve local dependency overrides, and it fried my brain... This was the easiest way for me to not be blocked by that anymore.

File: packages/jsii-pacmak/lib/targets/go/package.ts
Comment (RomainMuller) [reply packages/jsii-pacmak/lib/targets/go/package.ts:111]: Duh I always forget this is a thing... -_-

File: packages/jsii-pacmak/test/generated-code/__snapshots__/target-go.test.ts.snap
Comment (RomainMuller) [reply packages/jsii-pacmak/test/generated-code/__snapshots__/target-go.test.ts.snap:211]: Yes, it is. There is no way to get a type by name, because a type that isn't actually used (as in - being the type of some declared memory location) is not even present in the compiled program.

File: packages/@jsii/go-runtime/build-tools/gen-calc.ts
Comment (eladb) [reply packages/@jsii/go-runtime/build-tools/gen-calc.ts:18]: Local deps should work but I don't think it's a huge problem to use code-only here. Maybe use the fully qualified long name

File: packages/@jsii/go-runtime/jsii-runtime-go/type-registry.go
Comment (RomainMuller) [reply packages/@jsii/go-runtime/jsii-runtime-go/type-registry.go:107]: That's a good idea.

**Review by MrArnoldPalmer — APPROVED** [review]
👍🏻 love it @RomainMuller

**Pull request merged**
